### ISSUE TITLE
Additive Attention Layer is not really Additive Attention

### ISSUE THREAD

**Issue by AGalassi** [tensorflow/tensorflow#43471]
In Bahdanau's paper, after applying the non-linearity (tanh), there is a dot product between a vector va and the result of the non-linearity (page 12 of the paper). Indeed the result of the tanh(...) operation is a n x 1 matrix. va is a n x 1 matrix, which is transposed into a 1 x n. So the final result is a single score for each key. After this, the softmax function is applied to the set of scores.
Instead, the current AdditiveAttention layer does a simple (optional) scaling along the feature dimension.

Moreover, the score is computed incorrectly, as reported by #39332

**Comment by ravikyram** [comment]
@AGalassi 

Request you to fill [issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).
Please, share colab link or simple standalone code to eproduce the issue in our environment. It helps us in localizing the issue faster. Thanks!

**Comment by AGalassi** [comment]
> @AGalassi
> 
> Request you to fill [issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).
> Please, share colab link or simple standalone code to eproduce the issue in our environment. It helps us in localizing the issue faster. Thanks!

Thanks for the fast answer. It's not about the execution, it's the implementation itself (and the documentation).
The current implementation is some sort of different co-attention that does not correspond to Additive Attention nor to Bahdanau's paper.

The file is tensorflow/python/keras/layers/dense_attention.py, the class is AdditiveAttention. If you need more information of any sort just tell me.

**Comment by philyoun** [comment]
Any update to this issue? 
I was reviewing the source code of `keras.layers.AdditiveAttention()`, and found it was not implemented as the original paper as well.

**Comment by philyoun** [comment]
Upon reading #39332, it seems like I should aware of the base implementation of this source code and call additional this and that(matrices W1, W2 and so on) by my own when I use this..

**Comment by AGalassi** [comment]
As far as I know, no updates.
Also, I am not sure why @jvishnuvardhan has labeled this as a feature and not a bug. Maybe because is more of a content matter than a technical matter.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
